File: win/directx/sdk-9.0/Samples/C#/Direct3D/VertexBlend/vertexBlend.cs
//-----------------------------------------------------------------------------
// File: VertexBlend.cs
//
// Desc: Example code showing how to do a skinning effect, using the vertex
//       blending feature of Direct3D. Normally, Direct3D transforms each
//       vertex through the world matrix. The vertex blending feature,
//       however, uses mulitple world matrices and a per-vertex blend factor
//       to transform each vertex.
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;




namespace VertexBlendSample
{
	/// <summary>
	/// Application class. The base class (GraphicsSample) provides the generic 
	/// functionality needed in all Direct3D samples. MyGraphicsSample adds 
	/// functionality specific to this sample program.
	/// </summary>
	public class MyGraphicsSample : GraphicsSample
	{
        /// <summary>
        /// Custom vertex which includes a blending factor
        /// </summary>
		public struct BlendVertex
		{
			public Vector3 v;       // Referenced as v0 in the vertex shader
			public float blend;   // Referenced as v1.x in the vertex shader
			public Vector3 n;       // Referenced as v3 in the vertex shader
			public float tu, tv;  // Referenced as v7 in the vertex shader

			public static readonly VertexFormats Format = VertexFormats.PositionBlend1 | VertexFormats.Normal |  VertexFormats.Texture1;
		};

		private GraphicsFont drawingFont = null;                 // Font for drawing text
		private GraphicsMesh blendObject = null;           // Object to use for vertex blending
		private int numVertices = 0;
		private int numFaces = 0;
		private VertexBuffer vertexBuffer = null;    
		private IndexBuffer indexBuffer = null;

		private Matrix upperArm = Matrix.Identity;       // Vertex blending matrices
		private Matrix lowerArm = Matrix.Identity;

		private VertexShader shader = null;    // Vertex shader
		private VertexDeclaration declaration = null;
		private bool useShader = false;
		private System.Windows.Forms.MenuItem mnuOptions;
		private System.Windows.Forms.MenuItem mnuUseVS;


		

		/// <summary>
		/// Application constructor. Sets attributes for the app.
		/// </summary>
		public MyGraphicsSample()
		{
			// Set the window text
			this.Text = "VertexBlend: Surface Skinning Example";
            try
            {
                // Load the icon from our resources
                System.Resources.ResourceManager resources = new System.Resources.ResourceManager(this.GetType());
                this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            }
            catch
            {
                // It's no big deal if we can't load our icons, but try to load the embedded one
                try { this.Icon = new System.Drawing.Icon(this.GetType(), "directx.ico"); } 
                catch {}
            }

			drawingFont = new GraphicsFont("Arial", System.Drawing.FontStyle.Bold);
			enumerationSettings.AppUsesDepthBuffer = true;

			// Add our new menu options
			this.mnuOptions = new System.Windows.Forms.MenuItem();
			this.mnuUseVS = new System.Windows.Forms.MenuItem();
			// Add the Options menu to the main menu
			this.mnuMain.MenuItems.Add(this.mnuOptions);
			this.mnuOptions.Index = 1;
			this.mnuOptions.Text = "&Options";

			this.mnuOptions.MenuItems.Add(this.mnuUseVS);
			this.mnuUseVS.Text = "Use custom &vertex shader";
			this.mnuUseVS.Shortcut = System.Windows.Forms.Shortcut.CtrlV;
			this.mnuUseVS.ShowShortcut = true;
			this.mnuUseVS.Click += new System.EventHandler(this.UseCustomShaderClick);
		}




		/// <summary>
		/// Called when our menu item is clicked.
		/// </summary>
		private void UseCustomShaderClick(object sender, System.EventArgs e)
		{
			useShader = !useShader;
			mnuUseVS.Checked = !mnuUseVS.Checked;
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for animating the scene.
		/// </summary>
		protected override void FrameMove()
		{
			// Set the vertex blending matrices for this frame
			Vector3 vAxis = new Vector3(2 + (float)Math.Sin(appTime*3.1f), 2 + (float)Math.Sin(appTime*3.3f), (float)Math.Sin(appTime*3.5f)); 
			lowerArm = Matrix.RotationAxis(vAxis, (float)Math.Sin(3*appTime));
			upperArm =Matrix.Identity;

			// Set the vertex shader constants. Note: outside of the blend matrices,
			// most of these values don't change, so don't need to really be set every
			// frame. It's just done here for clarity
			if (useShader)
			{
				// Some basic constants
				Vector4 vZero = new Vector4(0,0,0,0);
				Vector4 vOne = new Vector4(1,1,1,1);

				// Lighting vector (normalized) and material colors. (Use non-yellow light
				// to show difference from non-vertex shader case.)
				Vector4 vLight = new Vector4(0.5f, 1.0f, -1.0f, 0.0f);
				vLight.Normalize();
				float[] fDiffuse = { 0.5f, 1.00f, 0.00f, 0.00f }; 
				float[] fAmbient = { 0.25f, 0.25f, 0.25f, 0.25f };

				// Vertex shader operations use transposed matrices
				Matrix matWorld0Transpose = new Matrix(), matWorld1Transpose = new Matrix();
				Matrix matView, matProj, matViewProj, matViewProjTranspose = new Matrix();
				matView = device.Transform.View;
				matProj = device.Transform.Projection;
				matViewProj = Matrix.Multiply(matView, matProj);
				matWorld0Transpose.Transpose(upperArm);
				matWorld1Transpose.Transpose(lowerArm);
				matViewProjTranspose.Transpose(matViewProj);

				// Set the vertex shader constants
				device.SetVertexShaderConstant(0, new Vector4[] { vZero });
				device.SetVertexShaderConstant(1, new Vector4[] { vOne });
				device.SetVertexShaderConstant(4, new Matrix[] { matWorld0Transpose });
				device.SetVertexShaderConstant(8, new Matrix[] { matWorld1Transpose });
				device.SetVertexShaderConstant(12, new Matrix[] { matViewProjTranspose });
				device.SetVertexShaderConstant(20, new Vector4[] { vLight });
				device.SetVertexShaderConstant(21, fDiffuse);
				device.SetVertexShaderConstant(22, fAmbient);
			}

		}




		/// <summary>
		/// Called once per frame, the call is the entry point for 3d rendering. This 
		/// function sets up render states, clears the viewport, and renders the scene.
		/// </summary>
		protected override void Render()
		{
			// Clear the viewport
			device.Clear(ClearFlags.Target | ClearFlags.ZBuffer, System.Drawing.Color.Blue, 1.0f, 0);

			device.BeginScene();

			if (useShader)        
			{
				device.VertexFormat = BlendVertex.Format;
				device.VertexShader = shader;
				device.SetStreamSource(0, vertexBuffer, 0, DXHelp.GetTypeSize(typeof(BlendVertex)));
				device.Indices = indexBuffer;
				device.DrawIndexedPrimitives(PrimitiveType.TriangleList, 0, 0, numVertices, 0, numFaces);
			}
			else
			{
				device.VertexShader = null;
				// Enable vertex blending using API
				device.Transform.World = upperArm;
				device.Transform.World1 = lowerArm;
				renderState.VertexBlend = VertexBlend.OneWeights;

				// Display the object
				blendObject.Render(device);
			}


			// Output statistics
			drawingFont.DrawText(2,  1, System.Drawing.Color.Yellow, frameStats);
			drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats);

			if (useShader)
				drawingFont.DrawText(2, 40, System.Drawing.Color.White, "Using vertex shader");
			else
				drawingFont.DrawText(2, 40, System.Drawing.Color.White, "Using RenderState.VertexBlend");

			device.EndScene();
		}




		/// <summary>
        /// The device has been created.  Resources that are not lost on
        /// Reset() can be created here -- resources in Pool.Managed,
        /// Pool.Scratch, or Pool.SystemMemory.  Image surfaces created via
        /// CreateImageSurface are never lost and can be created here.  Vertex
        /// shaders and pixel shaders can also be created here as they are not
        /// lost on Reset().
		/// </summary>
		protected override void InitializeDeviceObjects()
		{
			drawingFont.InitializeDeviceObjects(device);
			// Load an object to render
			try
			{
				if (blendObject == null)
					blendObject = new GraphicsMesh();

				blendObject.Create(device, "mslogo.x");
			}
			catch
			{
				SampleException e = new MediaNotFoundException();
				HandleSampleException(e, ApplicationMessage.ApplicationMustExit);
				throw e;
			}


			if ((BehaviorFlags.HardwareVertexProcessing || BehaviorFlags.MixedVertexProcessing) &&
				Caps.VertexShaderVersion.Major < 1)
			{
				// No VS available, so don't try to use it or allow user to
				// switch to it
				useShader = false;
				mnuUseVS.Enabled = false;
			}
			else if (Caps.MaxVertexBlendMatrices < 2)
			{
				// No blend matrices available, so don't try to use them or 
				// allow user to switch to them
				useShader = true;
				mnuUseVS.Enabled = false;
			}
			else
			{
				// Both techniques available, so default to blend matrices and 
				// allow the user to switch techniques
				useShader = false;
				mnuUseVS.Enabled = true;
			}

			// Set a custom FVF for the mesh
			blendObject.SetVertexFormat(device, BlendVertex.Format);

			// Add blending weights to the mesh
			// Gain acces to the mesh's vertices
			VertexBuffer vb = null;
			BlendVertex[] vertices = null;
			int numVertices = blendObject.SystemMesh.NumberVertices;
			vb = blendObject.SystemMesh.VertexBuffer;
			vertices = (BlendVertex[])vb.Lock(0,typeof(BlendVertex), 0, numVertices);

			// Calculate the min/max z values for all the vertices
			float fMinX =  1e10f;
			float fMaxX = -1e10f;

			for (int i=0; i<numVertices; i++)
			{
				if (vertices[i].v.X < fMinX) 
					fMinX = vertices[i].v.X;
				if (vertices[i].v.X > fMaxX) 
					fMaxX = vertices[i].v.X;
			}

			for (int i=0; i<numVertices; i++)
			{
				// Set the blend factors for the vertices
				float a = (vertices[i].v.X - fMinX) / (fMaxX - fMinX);
				vertices[i].blend = 1.0f-(float)Math.Sin(a*(float)Math.PI*1.0f);
			}

			// Done with the mesh's vertex buffer data
			vb.Unlock();
			vb.Dispose();
		}




		/// <summary>
        /// The device exists, but may have just been Reset().  Resources in
        /// Pool.Default and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
		/// </summary>
		protected override void RestoreDeviceObjects(System.Object sender, System.EventArgs e)
		{
			// Restore mesh's local memory objects
			blendObject.RestoreDeviceObjects(device, null);

			// Get access to the mesh vertex and index buffers
			vertexBuffer = blendObject.LocalMesh.VertexBuffer;
			indexBuffer = blendObject.LocalMesh.IndexBuffer;
			numVertices = blendObject.LocalMesh.NumberVertices;
			numFaces = blendObject.LocalMesh.NumberFaces;

			if (BehaviorFlags.SoftwareVertexProcessing ||
				Caps.VertexShaderVersion.Major >= 1)
			{
				// Setup the vertex declaration
				VertexElement[] decl = VertexInformation.DeclaratorFromFormat(BlendVertex.Format);
				declaration = new VertexDeclaration(device, decl);

				// Create vertex shader from a file
				try
				{
					shader = GraphicsUtility.CreateVertexShader(device, "blend.vsh");
				}
				catch
				{
					SampleException d3de = new MediaNotFoundException();
					HandleSampleException(d3de, ApplicationMessage.ApplicationMustExit);
					throw d3de;
				}
			}
			// Set miscellaneous render states
			renderState.ZBufferEnable = true;
			renderState.Ambient = System.Drawing.Color.FromArgb(0x00404040);

			// Set the projection matrix
			float fAspect = device.PresentationParameters.BackBufferWidth / (float)device.PresentationParameters.BackBufferHeight;
			device.Transform.Projection = Matrix.PerspectiveFovLH((float)Math.PI/4, fAspect, 1.0f, 10000.0f);

			// Set the app view matrix for normal viewing
			Vector3 vEyePt    = new Vector3(0.0f,-5.0f,-10.0f);
			Vector3 vLookatPt = new Vector3(0.0f, 0.0f,  0.0f);
			Vector3 vUpVec    = new Vector3(0.0f, 1.0f,  0.0f);
			device.Transform.View = Matrix.LookAtLH(vEyePt, vLookatPt, vUpVec);

			// Create a directional light. (Use yellow light to distinguish from
			// vertex shader case.)
			GraphicsUtility.InitLight(device.Lights[0], LightType.Directional, -0.5f, -1.0f, 1.0f);
			device.Lights[0].Diffuse = System.Drawing.Color.Yellow;
			device.Lights[0].Commit();
			device.Lights[0].Enabled = true;
			renderState.Lighting = true;
		}




        /// <summary>
        /// Invalidate the device objects
        /// </summary>
        protected override void InvalidateDeviceObjects(System.Object sender, System.EventArgs e)
		{
			if (shader != null)
				shader.Dispose();
		}




		/// <summary>
		/// Called when the app is exiting, or the device is being changed, this 
		/// function deletes any device-dependent objects.
		/// </summary>
		protected override void DeleteDeviceObjects(System.Object sender, System.EventArgs e)
		{
			blendObject.Dispose();
		}




		/// <summary>
		/// Called during device initialization, this code checks the device for some 
		/// minimum set of capabilities
		/// </summary>
		protected override bool ConfirmDevice(Caps caps, VertexProcessingType vertexProcessingType, 
			Format adapterFormat, Format backBufferFormat)
		{
			if (vertexProcessingType == VertexProcessingType.PureHardware)
				return false; // GetTransform doesn't work on PUREDEVICE

			// Check that the device supports at least one of the two techniques
			// used in this sample: either a vertex shader, or at least two blend
			// matrices and a directional light.

			if ((vertexProcessingType == VertexProcessingType.Hardware) ||
				(vertexProcessingType == VertexProcessingType.Mixed))
			{
				if (caps.VertexShaderVersion.Major >= 1)
					return true;
			}
			else
			{
				// Software vertex processing always supports vertex shaders
				return true;
			}

			// Check that the device can blend vertices with at least two matrices
			// (Software can always do up to 4 blend matrices)
			if (caps.MaxVertexBlendMatrices < 2)
				return false;

			// If this is a TnL device, make sure it supports directional lights
			if ((vertexProcessingType == VertexProcessingType.Hardware) ||
				(vertexProcessingType == VertexProcessingType.Mixed))
			{
				if (!caps.VertexProcessingCaps.SupportsDirectionAllLights)
					return false;
			}
			return true;
		}




        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main() 
        {
            using (MyGraphicsSample d3dApp = new MyGraphicsSample())
            {                                 
                if (d3dApp.CreateGraphicsSample())
                    d3dApp.Run();
            }
        }
    }
}

File: win/directx/sdk-9.0/Samples/C++/Direct3D/VertexBlend/vertexblend.cpp
//-----------------------------------------------------------------------------
// File: VertexBlend.cpp
//
// Desc: Example code showing how to do a skinning effect, using the vertex
//       blending feature of Direct3D. Normally, Direct3D transforms each
//       vertex through the world matrix. The vertex blending feature,
//       however, uses mulitple world matrices and a per-vertex blend factor
//       to transform each vertex.
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#define STRICT
#include <Windows.h>
#include <commctrl.h>
#include <math.h>
#include <stdio.h>
#include <D3DX9.h>
#include "DXUtil.h"
#include "D3DEnumeration.h"
#include "D3DSettings.h"
#include "D3DApp.h"
#include "D3DFile.h"
#include "D3DFont.h"
#include "D3DUtil.h"
#include "resource.h"




//-----------------------------------------------------------------------------
// Name: struct BLENDVERTEX
// Desc: Custom vertex which includes a blending factor
//-----------------------------------------------------------------------------
struct BLENDVERTEX
{
    D3DXVECTOR3 v;       // Referenced as v0 in the vertex shader
    FLOAT       blend;   // Referenced as v1.x in the vertex shader
    D3DXVECTOR3 n;       // Referenced as v3 in the vertex shader
    FLOAT       tu, tv;  // Referenced as v7 in the vertex shader

    static const DWORD FVF;
};
const DWORD BLENDVERTEX::FVF = D3DFVF_XYZB1 | D3DFVF_NORMAL | D3DFVF_TEX1;



//-----------------------------------------------------------------------------
// Name: class CMyD3DApplication
// Desc: Application class. The base class (CD3DApplication) provides the 
//       generic functionality needed in all Direct3D samples. CMyD3DApplication 
//       adds functionality specific to this sample program.
//-----------------------------------------------------------------------------
class CMyD3DApplication : public CD3DApplication
{
    CD3DFont*    m_pFont;
    CD3DMesh*    m_pObject;           // Object to use for vertex blending
    DWORD        m_dwNumVertices;
    DWORD        m_dwNumFaces;
    LPDIRECT3DVERTEXBUFFER9 m_pVB;    
    LPDIRECT3DINDEXBUFFER9  m_pIB;

    D3DXMATRIXA16   m_matUpperArm;       // Vertex blending matrices
    D3DXMATRIXA16   m_matLowerArm;

    LPDIRECT3DVERTEXSHADER9 m_pVertexShader;    // Vertex shader
    BOOL         m_bUseVertexShader;

protected:
    HRESULT OneTimeSceneInit();
    HRESULT InitDeviceObjects();
    HRESULT RestoreDeviceObjects();
    HRESULT InvalidateDeviceObjects();
    HRESULT DeleteDeviceObjects();
    HRESULT Render();
    HRESULT FrameMove();
    HRESULT FinalCleanup();
    HRESULT ConfirmDevice( D3DCAPS9*, DWORD, D3DFORMAT, D3DFORMAT );
    LRESULT MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam );

public:
    CMyD3DApplication();
};




//-----------------------------------------------------------------------------
// Name: WinMain()
// Desc: Entry point to the program. Initializes everything, and goes into a
//       message-processing loop. Idle time is used to render the scene.
//-----------------------------------------------------------------------------
INT WINAPI WinMain( HINSTANCE hInst, HINSTANCE, LPSTR, INT )
{
    CMyD3DApplication d3dApp;
    
    InitCommonControls();
    if( FAILED( d3dApp.Create( hInst ) ) )
        return 0;

    return d3dApp.Run();
}




//-----------------------------------------------------------------------------
// Name: CMyD3DApplication()
// Desc: Application constructor. Sets attributes for the app.
//-----------------------------------------------------------------------------
CMyD3DApplication::CMyD3DApplication()
{
    m_strWindowTitle    = _T("VertexBlend: Surface Skinning Example");
    m_d3dEnumeration.AppUsesDepthBuffer   = TRUE;

    m_pFont             = new CD3DFont( _T("Arial"), 12, D3DFONT_BOLD );
    m_pObject           = new CD3DMesh();
    m_pVB               = NULL;
    m_pIB               = NULL;
    m_pVertexShader     = NULL;
    m_bUseVertexShader  = FALSE;
}




//-----------------------------------------------------------------------------
// Name: OneTimeSceneInit()
// Desc: Called during initial app startup, this function performs all the
//       permanent initialization.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::OneTimeSceneInit()
{
    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FrameMove()
// Desc: Called once per frame, the call is the entry point for animating
//       the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FrameMove()
{
    // Set the vertex blending matrices for this frame
    D3DXVECTOR3 vAxis( 2+sinf(m_fTime*3.1f), 2+sinf(m_fTime*3.3f), sinf(m_fTime*3.5f) ); 
    D3DXMatrixRotationAxis( &m_matLowerArm, &vAxis, sinf(3*m_fTime) );
    D3DXMatrixIdentity( &m_matUpperArm );

    // Set the vertex shader constants. Note: outside of the blend matrices,
    // most of these values don't change, so don't need to really be set every
    // frame. It's just done here for clarity
    if( m_bUseVertexShader )
    {
        // Some basic constants
        D3DXVECTOR4 vZero(0,0,0,0);
        D3DXVECTOR4 vOne(1,1,1,1);

        // Lighting vector (normalized) and material colors. (Use red light
        // to show difference from non-vertex shader case.)
        D3DXVECTOR4 vLight( 0.5f, 1.0f, -1.0f, 0.0f );
        D3DXVec4Normalize( &vLight, &vLight );
        FLOAT       fDiffuse[] = { 1.00f, 1.00f, 0.00f, 0.00f };
        FLOAT       fAmbient[] = { 0.25f, 0.25f, 0.25f, 0.25f };

        // Vertex shader operations use transposed matrices
        D3DXMATRIXA16 matWorld0Transpose, matWorld1Transpose;
        D3DXMATRIXA16 matView, matProj, matViewProj, matViewProjTranspose;
        m_pd3dDevice->GetTransform( D3DTS_VIEW,       &matView );
        m_pd3dDevice->GetTransform( D3DTS_PROJECTION, &matProj );
        D3DXMatrixMultiply( &matViewProj, &matView, &matProj );
        D3DXMatrixTranspose( &matWorld0Transpose, &m_matUpperArm );
        D3DXMatrixTranspose( &matWorld1Transpose, &m_matLowerArm );
        D3DXMatrixTranspose( &matViewProjTranspose, &matViewProj );

        // Set the vertex shader constants
        m_pd3dDevice->SetVertexShaderConstantF(  0, (float*)&vZero,    1 );
        m_pd3dDevice->SetVertexShaderConstantF(  1, (float*)&vOne,     1 );
        m_pd3dDevice->SetVertexShaderConstantF(  4, (float*)&matWorld0Transpose,   4 );
        m_pd3dDevice->SetVertexShaderConstantF(  8, (float*)&matWorld1Transpose,   4 );
        m_pd3dDevice->SetVertexShaderConstantF( 12, (float*)&matViewProjTranspose, 4 );
        m_pd3dDevice->SetVertexShaderConstantF( 20, (float*)&vLight,   1 );
        m_pd3dDevice->SetVertexShaderConstantF( 21, (float*)&fDiffuse, 1 );
        m_pd3dDevice->SetVertexShaderConstantF( 22, (float*)&fAmbient, 1 );
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: Render()
// Desc: Called once per frame, the call is the entry point for 3d
//       rendering. This function sets up render states, clears the
//       viewport, and renders the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::Render()
{
    // Clear the backbuffer
    m_pd3dDevice->Clear( 0L, NULL, D3DCLEAR_TARGET|D3DCLEAR_ZBUFFER, 
                         0x000000ff, 1.0f, 0L );

    // Begin the scene
    if( SUCCEEDED( m_pd3dDevice->BeginScene() ) )
    {
        if( m_bUseVertexShader )        
        {
            m_pd3dDevice->SetFVF( BLENDVERTEX::FVF );
            m_pd3dDevice->SetVertexShader( m_pVertexShader );
            m_pd3dDevice->SetStreamSource( 0, m_pVB, 0, sizeof(BLENDVERTEX) );
            m_pd3dDevice->SetIndices( m_pIB );
            m_pd3dDevice->DrawIndexedPrimitive( D3DPT_TRIANGLELIST, 0, 0, m_dwNumVertices,
                                                0, m_dwNumFaces );
        }
        else
        {
            // Enable vertex blending using API
            m_pd3dDevice->SetVertexShader( NULL );
            m_pd3dDevice->SetTransform( D3DTS_WORLD,  &m_matUpperArm );
            m_pd3dDevice->SetTransform( D3DTS_WORLD1, &m_matLowerArm );
            m_pd3dDevice->SetRenderState( D3DRS_VERTEXBLEND, D3DVBF_1WEIGHTS );

            // Display the object
            m_pObject->Render( m_pd3dDevice );
        }

        // Output statistics
        m_pFont->DrawText( 2,  0, D3DCOLOR_ARGB(255,255,255,0), m_strFrameStats );
        m_pFont->DrawText( 2, 20, D3DCOLOR_ARGB(255,255,255,0), m_strDeviceStats );

        if( m_bUseVertexShader )
            m_pFont->DrawText( 2, 40, D3DCOLOR_ARGB(255,255,255,255), _T("Using vertex shader") );
        else
            m_pFont->DrawText( 2, 40, D3DCOLOR_ARGB(255,255,255,255), _T("Using D3DRS_VERTEXBLEND") );

        // End the scene.
        m_pd3dDevice->EndScene();
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: InitDeviceObjects()
// Desc: Initialize scene objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InitDeviceObjects()
{
    HRESULT hr;

    // Initialize the font's internal textures
    m_pFont->InitDeviceObjects( m_pd3dDevice );

    // Load an object to render
    if( FAILED( m_pObject->Create( m_pd3dDevice, _T("mslogo.x") ) ) )
        return D3DAPPERR_MEDIANOTFOUND;

    if( ( ( m_dwCreateFlags & D3DCREATE_HARDWARE_VERTEXPROCESSING ) ||
          ( m_dwCreateFlags & D3DCREATE_MIXED_VERTEXPROCESSING ) ) &&
        m_d3dCaps.VertexShaderVersion < D3DVS_VERSION(1,0) )
    {
        // No VS available, so don't try to use it or allow user to
        // switch to it
        m_bUseVertexShader = FALSE;
        EnableMenuItem( GetMenu( m_hWnd ), IDM_USEVERTEXSHADER, MF_GRAYED );
    }
    else if( m_d3dCaps.MaxVertexBlendMatrices < 2 )
    {
        // No blend matrices available, so don't try to use them or 
        // allow user to switch to them
        m_bUseVertexShader = TRUE;
        EnableMenuItem( GetMenu( m_hWnd ), IDM_USEVERTEXSHADER, MF_GRAYED );
    }
    else
    {
        // Both techniques available, so default to blend matrices and 
        // allow the user to switch techniques
        m_bUseVertexShader = FALSE;
        EnableMenuItem( GetMenu( m_hWnd ), IDM_USEVERTEXSHADER, MF_ENABLED );
    }

    // Set a custom FVF for the mesh
    m_pObject->SetFVF( m_pd3dDevice, BLENDVERTEX::FVF );

    // Add blending weights to the mesh
    {
        // Gain acces to the mesh's vertices
        LPDIRECT3DVERTEXBUFFER9 pVB;
        BLENDVERTEX* pVertices;
        DWORD        dwNumVertices = m_pObject->GetSysMemMesh()->GetNumVertices();
        m_pObject->GetSysMemMesh()->GetVertexBuffer( &pVB );
        pVB->Lock( 0, 0, (void**)&pVertices, 0 );

        // Calculate the min/max z values for all the vertices
        FLOAT fMinX =  1e10f;
        FLOAT fMaxX = -1e10f;

        for( DWORD i=0; i<dwNumVertices; i++ )
        {
            if( pVertices[i].v.x < fMinX ) 
                fMinX = pVertices[i].v.x;
            if( pVertices[i].v.x > fMaxX ) 
                fMaxX = pVertices[i].v.x;
        }

        for( i=0; i<dwNumVertices; i++ )
        {
            // Set the blend factors for the vertices
            FLOAT a = ( pVertices[i].v.x - fMinX ) / ( fMaxX - fMinX );
            pVertices[i].blend = 1.0f-sinf(a*D3DX_PI*1.0f);
        }

        // Done with the mesh's vertex buffer data
        pVB->Unlock();
        pVB->Release();
    }

    if( ( m_dwCreateFlags & D3DCREATE_SOFTWARE_VERTEXPROCESSING ) ||
        m_d3dCaps.VertexShaderVersion >= D3DVS_VERSION(1,0) )
    {
        TCHAR        strVertexShaderPath[512];
        LPD3DXBUFFER pCode;

        // Find the vertex shader file
        if( FAILED( hr = DXUtil_FindMediaFileCb( strVertexShaderPath, 
            sizeof(strVertexShaderPath), _T("Blend.vsh") ) ) )
        { 
            return hr;
        }

        // Assemble the vertex shader from the file
        if( FAILED( hr = D3DXAssembleShaderFromFile( strVertexShaderPath, 
                                                     NULL, NULL, 0, &pCode, NULL ) ) )
        {
            return hr;
        }

        // Create the vertex shader
        hr = m_pd3dDevice->CreateVertexShader( (DWORD*)pCode->GetBufferPointer(),
                                               &m_pVertexShader );
        pCode->Release();
        if( FAILED( hr ) )
            return hr;
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: RestoreDeviceObjects()
// Desc: Restore device-memory objects and state after a device is created or
//       resized.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::RestoreDeviceObjects()
{
    m_pFont->RestoreDeviceObjects();

    // Restore mesh's local memory objects
    m_pObject->RestoreDeviceObjects( m_pd3dDevice );

    // Get access to the mesh vertex and index buffers
    m_pObject->GetLocalMesh()->GetVertexBuffer( &m_pVB );
    m_pObject->GetLocalMesh()->GetIndexBuffer( &m_pIB );
    m_dwNumVertices = m_pObject->GetLocalMesh()->GetNumVertices();
    m_dwNumFaces    = m_pObject->GetLocalMesh()->GetNumFaces();

    // Set miscellaneous render states
    m_pd3dDevice->SetRenderState( D3DRS_ZENABLE, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_AMBIENT, 0x00404040 );

    // Set the projection matrix
    D3DXMATRIXA16 matProj;
    FLOAT fAspect = m_d3dsdBackBuffer.Width / (FLOAT)m_d3dsdBackBuffer.Height;
    D3DXMatrixPerspectiveFovLH( &matProj, D3DX_PI/4, fAspect, 1.0f, 10000.0f );
    m_pd3dDevice->SetTransform( D3DTS_PROJECTION, &matProj );

    // Set the app view matrix for normal viewing
    D3DXVECTOR3 vEyePt    = D3DXVECTOR3( 0.0f,-5.0f,-10.0f );
    D3DXVECTOR3 vLookatPt = D3DXVECTOR3( 0.0f, 0.0f,  0.0f );
    D3DXVECTOR3 vUpVec    = D3DXVECTOR3( 0.0f, 1.0f,  0.0f );
    D3DXMATRIXA16 matView;
    D3DXMatrixLookAtLH( &matView, &vEyePt, &vLookatPt, &vUpVec );
    m_pd3dDevice->SetTransform( D3DTS_VIEW, &matView );

    // Create a directional light. (Use yellow light to distinguish from
    // vertex shader case.)
    D3DLIGHT9 light;
    D3DUtil_InitLight( light, D3DLIGHT_DIRECTIONAL, -0.5f, -1.0f, 1.0f );
    light.Diffuse.r = 1.0f;
    light.Diffuse.g = 1.0f;
    light.Diffuse.b = 0.0f;
    m_pd3dDevice->SetLight( 0, &light );
    m_pd3dDevice->LightEnable( 0, TRUE );
    m_pd3dDevice->SetRenderState( D3DRS_LIGHTING, TRUE );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: InvalidateDeviceObjects()
// Desc: Called when the device-dependent objects are about to be lost.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InvalidateDeviceObjects()
{
    m_pFont->InvalidateDeviceObjects();
    m_pObject->InvalidateDeviceObjects();

    SAFE_RELEASE( m_pVB );
    SAFE_RELEASE( m_pIB );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: DeleteDeviceObjects()
// Desc: Called when the app is exiting, or the device is being changed,
//       this function deletes any device dependent objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::DeleteDeviceObjects()
{
    m_pFont->DeleteDeviceObjects();
    m_pObject->Destroy();

    SAFE_RELEASE( m_pVertexShader );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FinalCleanup()
// Desc: Called before the app exits, this function gives the app the chance
//       to cleanup after itself.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FinalCleanup()
{
    SAFE_DELETE( m_pFont );
    SAFE_DELETE( m_pObject );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: ConfirmDevice()
// Desc: Called during device initialization, this code checks the device
//       for some minimum set of capabilities
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::ConfirmDevice( D3DCAPS9* pCaps, DWORD dwBehavior, 
                                          D3DFORMAT adapterFormat, D3DFORMAT backBufferFormat )
{
    if( dwBehavior & D3DCREATE_PUREDEVICE )
        return E_FAIL; // GetTransform doesn't work on PUREDEVICE

    // Check that the device supports at least one of the two techniques
    // used in this sample: either a vertex shader, or at least two blend
    // matrices and a directional light.

    if( (dwBehavior & D3DCREATE_HARDWARE_VERTEXPROCESSING ) ||
        (dwBehavior & D3DCREATE_MIXED_VERTEXPROCESSING ) )
    {
        if( pCaps->VertexShaderVersion >= D3DVS_VERSION(1,0) )
            return S_OK;
    }
    else
    {
        // Software vertex processing always supports vertex shaders
        return S_OK;
    }

    // Check that the device can blend vertices with at least two matrices
    // (Software can always do up to 4 blend matrices)
    if( pCaps->MaxVertexBlendMatrices < 2 )
        return E_FAIL;

    // If this is a TnL device, make sure it supports directional lights
    if( (dwBehavior & D3DCREATE_HARDWARE_VERTEXPROCESSING ) ||
        (dwBehavior & D3DCREATE_MIXED_VERTEXPROCESSING ) )
    {
        if( !(pCaps->VertexProcessingCaps & D3DVTXPCAPS_DIRECTIONALLIGHTS ) )
            return E_FAIL;
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: MsgProc()
// Desc: Message proc function to handle key and menu input
//-----------------------------------------------------------------------------
LRESULT CMyD3DApplication::MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam,
                                    LPARAM lParam )
{
    // Handle menu commands
    if( WM_COMMAND == uMsg )
    {
        switch( LOWORD(wParam) )
        {
            case IDM_USEVERTEXSHADER:
                m_bUseVertexShader = !m_bUseVertexShader;
                CheckMenuItem( GetMenu(hWnd), IDM_USEVERTEXSHADER,
                               m_bUseVertexShader ? MF_CHECKED : MF_UNCHECKED );
                break;
        }
    }

    // Pass remaining messages to default handler
    return CD3DApplication::MsgProc( hWnd, uMsg, wParam, lParam );
}



